packages/live-share-media/src/VolumeLimiter.ts

```diff
@@ -83,100 +88,85 @@ export class VolumeLimiter {
-                        this._player.volume = volume;
-                        this._inRamp = true;
-                    } else if (this._inRamp) {
-                        // We're at the end of a ramp
-                        volume = targetVolume;
-                        this._player.volume = volume;
-                        this._inRamp = false;
-                    }
-                }
-
-                // Schedule next animation frame
+            if (this.milliIntoRamp() <= this._rampDuration.milliseconds) {
+                const newVolume = this.computeRampVolume();
+                console.log("adjusting", newVolume);
+                this._player.volume = newVolume;
                 this.scheduleAnimationFrame(adjustVolume);
             } else {
-                // Reset volume to original level
-                this._player.volume = this._startVolume;
+                const newVolume = this.computeTargetVolume();
+                console.log("adjusting", newVolume);
+                this._player.volume = newVolume;
                 this._running = false;
-                this._endTime = 0;
-                this._startTime = 0;
-                this._startVolume = 0;
             }
-        };
+        }
 
-        this._running = true;
         this._startTime = new Date().getTime();
         this._startVolume = this._player.volume;
-        this.scheduleAnimationFrame(adjustVolume)
+        if (!this._running) {
+            this._running = true
+            this.scheduleAnimationFrame(adjustVolume);
+        }
     }
 
     private scheduleAnimationFrame(callback: FrameRequestCallback): void {
-        if (requestAnimationFrame) {
+        if (typeof requestAnimationFrame == "function") {
             requestAnimationFrame(callback);
         } else {
             setTimeout(callback, 20);
         }
     }
 
+    private computeRampVolume(): number {
+        const volumeDifference = this.computeTargetVolume() - this._startVolume
+        const adjustmentFromStart = volumeDifference / this._rampDuration.milliseconds * this.milliIntoRamp()
+        return this._startVolume + adjustmentFromStart;
+    }
+
     private computeTargetVolume(): number {
-        switch (this._levelType) {
-            case LevelType.percentage:
-                return this._startVolume * this._level;
-            case LevelType.fixed:
-            default:
+        if (this._levelType == LevelType.percentage) {
+            if (this._limited) {
+                return this._selectedVolume * this._level;
+            } else {
+                return this._selectedVolume;
+            }
+        } else {
+            if (this._limited && this._selectedVolume > this._level) {
                 return this._level;
+            } else {
+                return this._selectedVolume;
+            }
         }
     }
-}
\ No newline at end of file
+
+    private milliIntoRamp(): number {
```

Comment:
Not sure I understand the meaning. How about 'startTimeToNow" or `setStartTime` or something?

Reply:
changed to `milliAfterRampStart`. How is that?